runtime/src/mm/cpp/InitializationScheme.cpp

```diff
@@ -0,0 +1,74 @@
+/*
+ * Copyright 2010-2021 JetBrains s.r.o. Use of this source code is governed by the Apache 2.0 license
+ * that can be found in the LICENSE file.
+ */
+
+#include "InitializationScheme.hpp"
+
+#include "Common.h"
+#include "ObjectOps.hpp"
+#include "ThreadData.hpp"
+
+using namespace kotlin;
+
+OBJ_GETTER(mm::InitThreadLocalSingleton, ThreadData* threadData, ObjHeader** location, const TypeInfo* typeInfo, void (*ctor)(ObjHeader*)) {
+    if (auto* value = *location) {
+        // Initialized by someone else.
+        RETURN_OBJ(value);
+    }
+    auto* value = mm::AllocateObject(threadData, typeInfo, OBJ_RESULT);
+    mm::SetHeapRef(location, value);
+#if KONAN_NO_EXCEPTIONS
+    ctor(value);
+#else
+    try {
+        ctor(value);
+    } catch (...) {
+        mm::SetStackRef(OBJ_RESULT, nullptr);
+        mm::SetHeapRef(location, nullptr);
+        throw;
+    }
+#endif
+    return value;
+}
+
+OBJ_GETTER(mm::InitSingleton, ThreadData* threadData, ObjHeader** location, const TypeInfo* typeInfo, void (*ctor)(ObjHeader*)) {
+    auto& initializingSingletons = threadData->initializingSingletons();
+
+    // Search from the top of the stack.
+    for (auto it = initializingSingletons.rbegin(); it != initializingSingletons.rend(); ++it) {
+        if (it->first == location) {
+            RETURN_OBJ(it->second);
+        }
+    }
+
+    ObjHeader* initializing = reinterpret_cast<ObjHeader*>(1);
+
+    // Spin lock.
+    ObjHeader* value = nullptr;
+    while ((value = __sync_val_compare_and_swap(location, nullptr, initializing)) == initializing) {
+    }
+    if (value != nullptr) {
+        // Initialized by someone else.
+        RETURN_OBJ(value);
+    }
+    auto* object = mm::AllocateObject(threadData, typeInfo, OBJ_RESULT);
+    initializingSingletons.push_back(std::make_pair(location, object));
```

Comment:
I have two general questions here:

1. Is this correct that in the ideal world the allocator behind `KStdVector` should trigger GC when there is no free memory to increase the vector capacity?
2. When do we zero reference fields of a newly allocated object? It seems that I missed something but I don't see such zeroing netiher in the `mm::AllocateObject` call chain nor in the bitcode of the generated ctor.

Reply:
> Is this correct that in the ideal world the allocator behind KStdVector should trigger GC when there is no free memory to increase the vector capacity?

I'm not sure. On one hand it makes sense, on the other hand it means that every memory allocation in runtime is also a safepoint, for which our runtime is certainly not ready. In the ideal world we should probably keep all our "system" data in a separate arena with enough preallocated memory, so that we (practically) never run into OOM with "system" allocations between GC sweeps.

Reply:
> When do we zero reference fields of a newly allocated object? It seems that I missed something but I don't see such zeroing netiher in the mm::AllocateObject call chain nor in the bitcode of the generated ctor.

We allocate zeroed memory: we call `calloc`. But I should've written a test on it for the object allocation. Also, there's a sanitizer, that would've fired if we were to access uninitialized data - MSAN. Unfortunately, it requires that all dependencies are built with it (even libc), and so would require us building yet another sysroot.

Reply:
> keep all our "system" data in a separate arena with enough preallocated memory, so that we (practically) never run into OOM with "system" allocations between GC sweeps.

Ok, you're probably right.

> We allocate zeroed memory: we call `calloc`

Thank you, I've just missed that `calloc` zeros memory by itself.